File: 2021/ts/six/main.ts
import { solve, test } from "../lib/utils.ts";

async function run() {
  const day = "six";
  await test(day, 1, partOne, 5934);
  await solve(day, 2, partOne, 388419);
  await test(day, 1, partTwo, 26984457539);
  await solve(day, 2, partTwo, 1740449478328);
}

function runSimulation(fish: Array<number>, duration: number): Map<number, number> {
  let fishAges = new Map<number, number>();
  fish.forEach((x) => {
    const age = fishAges.get(x) ?? 0;
      fishAges.set(x, age + 1);
  });

  for (let day = 0; day < duration; day++) {
    const tempFishAges = new Map<number, number>();
    let newFish = 0;
    fishAges.forEach((count, age) => {
      if (age === 0) {
        tempFishAges.set(6, count);
        newFish = count;
      } else {
        const existingYoungerCount = tempFishAges.get(age - 1);
        if (existingYoungerCount) {
          count += existingYoungerCount;
        }
        tempFishAges.set(age - 1, count);
      }
    });
    if (newFish > 0) {
      tempFishAges.set(8, newFish);
    }
    fishAges = tempFishAges;
  }
  return fishAges
}

function countFish(fishAges: Map<number, number>): number {
  let sum = 0;
  fishAges.forEach((v) => (sum += v));
  return sum;
}

function partOne(input: Array<string>): number {
  const initialPopulation = input[0].split(",").map((x) => parseInt(x, 10));
  const fishAges = runSimulation(initialPopulation, 80);
  return countFish(fishAges)
}

function partTwo(input: Array<string>): number {
  const initialPopulation = input[0].split(",").map((x) => parseInt(x, 10));
  const fishAges = runSimulation(initialPopulation, 256);
  return countFish(fishAges)
}

await run();

File: 2021/rust/six/src/main.rs
use std::collections::BTreeMap;
use std::time::Instant;

fn solve_part(solver: &dyn Fn(std::str::Lines) -> i64, expected: i64) {
    let lines = include_str!("../../../input/six.txt").lines();
    run_part(lines, solver, expected)
}

fn test_part(solver: &dyn Fn(std::str::Lines) -> i64, expected: i64) {
    let lines = include_str!("../../../input/test/six.txt").lines();
    run_part(lines, solver, expected)
}

fn run_part(lines: std::str::Lines, solver: &dyn Fn(std::str::Lines) -> i64, expected: i64) {
    let start = Instant::now();
    let answer = solver(lines.clone());
    let duration = start.elapsed();
    println!(
        "Part one answer: {}, expected {}. Completed in: {:?}",
        answer, expected, duration
    );
    assert_eq!(answer, expected);
}

fn main() {
    test_part(&part_one, 5934);
    solve_part(&part_one, 388419);

    test_part(&part_two, 26984457539);
    solve_part(&part_two, 1740449478328);
}

fn create_initial_population(mut lines: std::str::Lines) -> Vec<u8> {
    lines
        .next()
        .unwrap()
        .split(',')
        .map(|x| x.parse::<u8>().unwrap())
        .collect::<Vec<_>>()
}

fn run_sim(fish: Vec<u8>, duration: u16) -> BTreeMap<u8, u64> {
    let mut ages = BTreeMap::new();
    for f in fish.iter() {
        let age = f.to_owned();
        match ages.get(f) {
            Some(&x) => ages.insert(age, x + 1_u64),
            None => ages.insert(age, 1),
        };
    }

    for _n in 0..duration {
        let mut temp_ages = BTreeMap::new();
        let mut new_fish: u64 = 0;
        for (&age, &count) in ages.iter() {
            if age == 0 {
                temp_ages.insert(6, count);
                new_fish = count as u64;
            } else {
                let existing_age = temp_ages.get(&(age - 1));
                let mut temp_count = count;
                if let Some(x) = existing_age {
                    temp_count += x
                }
                temp_ages.insert(age - 1, temp_count);
            }
        }
        if new_fish > 0 {
            temp_ages.insert(8, new_fish);
        }
        ages = temp_ages;
    }
    ages
}

fn count_fish(ages: BTreeMap<u8, u64>) -> i64 {
    ages.values().map(|&x| x as i64).sum()
}

fn part_one(lines: std::str::Lines) -> i64 {
    let initial_population = create_initial_population(lines);
    let fish_ages = run_sim(initial_population, 80);
    count_fish(fish_ages)
}

fn part_two(lines: std::str::Lines) -> i64 {
    let initial_population = create_initial_population(lines);
    let fish_ages = run_sim(initial_population, 256);
    count_fish(fish_ages)
}
